src/domain/synthetics/trade/useTradeboxState.ts

```diff
@@ -48,38 +52,88 @@ type StoredTradeOptions = {
       short: string;
     };
   };
-  collateralAddress?: string;
+  collaterals?: {
+    [marketAddress: string]: {
+      long?: string;
+      short?: string;
+    };
+  };
 };
 
 const INITIAL_SYNTHETICS_TRADE_OPTIONS_STATE: StoredTradeOptions = {
   tradeType: TradeType.Long,
   tradeMode: TradeMode.Market,
   tokens: {},
   markets: {},
-  collateralAddress: undefined,
+  collaterals: {},
 };
 
 export function useTradeboxState(
   chainId: number,
   p: {
     marketsInfoData?: MarketsInfoData;
+    positionsInfoData?: PositionsInfoData;
     tokensData?: TokensData;
   }
 ) {
-  const { marketsInfoData, tokensData } = p;
+  const { marketsInfoData, tokensData, positionsInfoData } = p;
 
   const availableTokensOptions = useAvailableTokenOptions(chainId, { marketsInfoData, tokensData });
 
+  //#region Manual useMemo zone begin
+  // This ensures almost no reinitialization of all setters
+  /* eslint-disable react-hooks/exhaustive-deps */
+  const unstableRefAvailableSwapTokensAddresses = availableTokensOptions.swapTokens.map((t) => t.address);
+
+  const availableSwapTokenAddresses = useMemo(() => {
+    return unstableRefAvailableSwapTokensAddresses;
+  }, [unstableRefAvailableSwapTokensAddresses.sort().join(",")]);
+
+  const unstableRefAvailableIndexTokensAddresses = availableTokensOptions.indexTokens.map((t) => t.address);
+
+  const availableIndexTokensAddresses = useMemo(() => {
+    return unstableRefAvailableIndexTokensAddresses;
+  }, [unstableRefAvailableIndexTokensAddresses.sort().join(",")]);
+
+  const unstableRefAvailableMarketAddressIndexTokenMap =
+    availableTokensOptions.sortedAllMarkets &&
+    mapValues(
+      keyBy(availableTokensOptions.sortedAllMarkets, (market) => market.marketTokenAddress),
+      (market) => market.indexToken.address
+    );
+
+  const marketAddressIndexTokenMap: {
+    [marketAddress: string]: string;
+  } = useMemo(() => {
+    return unstableRefAvailableMarketAddressIndexTokenMap || EMPTY_OBJECT;
+  }, [JSON.stringify(unstableRefAvailableMarketAddressIndexTokenMap)]);
+
+  const unstableRefStrippedMarketInfo = mapValues(marketsInfoData || {}, (info) => ({
+    longTokenAddress: info.longTokenAddress,
+    shortTokenAddress: info.shortTokenAddress,
+    isLongTokenStable: Boolean(info.longToken.isStable),
+    isShortTokenStable: Boolean(info.shortToken.isStable),
+  }));
+
+  const strippedMarketInfo = useMemo(() => {
+    return unstableRefStrippedMarketInfo;
+  }, [JSON.stringify(unstableRefStrippedMarketInfo)]);
+  /* eslint-enable react-hooks/exhaustive-deps */
+  //#endregion Manual useMemo zone end
```

Comment:
Be no afraid, i just moved it and changed `strippedMarketInfo` and `marketAddressIndexTokenMap`